From 0a1fcec2408e281a10d65bfd9271d7b99bb357cd Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 207/296] ExifInterface: fix incorrect thumbnailLength in
 handleThumbnailFromJfif

Incorrect thumbnailLength in ExifInterface::handleThumbnailFromJfif
potentially leads to incomplete bytes of thumbnail.

The statement below potentially leads to incorrect thumbnailLength
caused by in.available.

thumbnailLength = Math.min(thumbnailLength, in.available() - thumbnailOffset);

Test: Invoke ExifInterface.getThumbnail to get thumbnail data stored in jpeg file.
---
 media/java/android/media/ExifInterface.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/media/java/android/media/ExifInterface.java b/media/java/android/media/ExifInterface.java
index 37ffa5f37a8..77d71f9d8e9 100755
--- a/media/java/android/media/ExifInterface.java
+++ b/media/java/android/media/ExifInterface.java
@@ -3203,7 +3203,7 @@ public class ExifInterface {
             int thumbnailLength = jpegInterchangeFormatLengthAttribute.getIntValue(mExifByteOrder);
 
             // The following code limits the size of thumbnail size not to overflow EXIF data area.
-            thumbnailLength = Math.min(thumbnailLength, in.available() - thumbnailOffset);
+            thumbnailLength = Math.min(thumbnailLength, in.mLength - thumbnailOffset);
             if (mMimeType == IMAGE_TYPE_JPEG || mMimeType == IMAGE_TYPE_RAF
                     || mMimeType == IMAGE_TYPE_RW2) {
                 thumbnailOffset += mExifOffset;
-- 
2.11.0

